pkg/cert/v1beta1/generate/generate.go

```diff
@@ -0,0 +1,262 @@
+		BasicConstraintsValid: true,
+		IsCA:                  true,
+	}
+
+	var err error
+	caKeyPair.key, err = rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		return err
+	}
+	der, err := x509.CreateCertificate(rand.Reader, template, template, caKeyPair.key.Public(), caKeyPair.key)
+	if err != nil {
+		return err
+	}
+	if err = caKeyPair.encode(x509.MarshalPKCS1PrivateKey(caKeyPair.key), der); err != nil {
+		return err
+	}
+	if caKeyPair.cert, err = x509.ParseCertificate(der); err != nil {
+		return err
+	}
+	klog.Info("Generate the self-signed CA certificate and private key.")
+	return nil
+}
+
+func (o *generateOptions) createCert(caKeyPair *certificates, keyPair *certificates) error {
+	now := time.Now()
+	template := &x509.Certificate{
+		SerialNumber: big.NewInt(1),
+		Subject: pkix.Name{
+			CommonName: o.fullServiceDomain,
+		},
+		DNSNames: []string{
+			consts.Service,
+			strings.Join([]string{consts.Service, o.namespace}, "."),
+			o.fullServiceDomain,
+		},
+		NotBefore:             now,
+		NotAfter:              now.Add(24 * time.Hour * 365 * 10),
+		KeyUsage:              x509.KeyUsageDigitalSignature | x509.KeyUsageKeyEncipherment,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+		BasicConstraintsValid: false,
+	}
+
+	var err error
+	keyPair.key, err = rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		return err
+	}
+	der, err := x509.CreateCertificate(rand.Reader, template, caKeyPair.cert, keyPair.key.Public(), caKeyPair.key)
+	if err != nil {
+		return err
+	}
+	if err = keyPair.encode(x509.MarshalPKCS1PrivateKey(keyPair.key), der); err != nil {
+		return err
+	}
+	klog.Info("Generate public certificate and private key signed with self-singed CA cert and private key.")
+	return nil
+}
+
+// createWebhookCertSecret create Secret embedded ca.key, ca.crt, tls.key and tls.cert
+func (o *generateOptions) createWebhookCertSecret(ctx context.Context, kubeClient client.Client, caKeyPair *certificates, keyPair *certificates) error {
+
+	certGeneratorJob := &batchv1.Job{}
+	if err := kubeClient.Get(ctx, client.ObjectKey{Namespace: o.namespace, Name: consts.JobName}, certGeneratorJob); err != nil {
+		return err
+	}
+
+	// Create secret with CA cert and server cert/key.
+	// Add ownerReferences to clean-up secret with cert generator Job.
+	jobUID := certGeneratorJob.UID
+	webhookCertSecret := &corev1.Secret{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Secret",
+			APIVersion: "v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      consts.Secret,
+			Namespace: o.namespace,
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion: "batch/v1",
+					Kind:       "Job",
+					Controller: pointer.BoolPtr(true),
+					Name:       consts.JobName,
+					UID:        jobUID,
+				},
+			},
+		},
+		Type: corev1.SecretTypeTLS,
+		Data: map[string][]byte{
+			"ca.key":  caKeyPair.keyPem,
+			"ca.crt":  caKeyPair.certPem,
```

Comment:
Why previously we added only `tls.key` and `tls.crt`: https://github.com/kubeflow/katib/blob/master/hack/cert-generator.sh#L150-L151 ?

Reply:
In the previous implementation, katib-cert-generator(KCG) creates a CertificateSigningRequest Resource (kubeCSR) containing a CSR and approves the kubeCSR using `kubectl approve`. Then the signing controller will sign it with the CA certificate and add the certificate to the status.certificate field of the kubeCSR. Finally, the webhook server functions by registering the CA certificate in FooWebhookConfiguration. 

In this implementation, KCG creates a CA certificate to sign the server certificate (`ca.crt` & `ca.key`) to use signing the server certificate (`tls.crt` & `tls.key`). Then, the webhook server functions by registering the first CA certificate to BarWebhookConfiguration.

Therefore, we previously added only `tls.crt` and `tls.key` to Secret.

Thanks for reading my explanation.

Reply:
Thank you for the explanation @tenzen-y!